fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Headers.kt

```diff
@@ -0,0 +1,225 @@
+package com.github.kittinunf.fuel.core
+
+data class HeaderName(val name: String) {
+    private val normalized = name.toUpperCase()
+    override fun hashCode(): Int {
+        return normalized.hashCode()
+    }
+
+    override fun equals(other: Any?): Boolean {
+        return (other is HeaderName) && other.normalized == normalized
+    }
+
+    override fun toString(): String {
+        return name
+    }
+}
+
+typealias HeaderValues = Collection<String>
+
+class Headers : MutableMap<String, HeaderValues> {
```

Comment:
Since most of the implementation is passing through to underlying hashMap contents, shall we use constructor delegation?

``` Kotlin
class Header(private val contents: HashMap<HeaderName, HeaderValues>) : MutableMap<HeaderName, HeaderValues> by contents {
  // only override something that is custom for example remove or some other
}
```

Reply:
@kittinunf I want the headers to be accessible by any `String` and not force users to wrap in `HeaderName`, so unfortunately that won't work.

Down the line, we probably want `Fuel` to be more immutable (and thus threadsafe) and change this to `Map<HeaderName, HeaderValues>` anyway, which will require MORE custom implementation.

Reply:
That's unfortunate ...

Reply:
Come to think about this by `users`, do you mean users who use Request class or the Request class itself? If it is the former, we can actually still expose only the API that interacts with `String` but internally (in Request class) we can do the wrapping in implementation? I would say if the wrapping `HeaderName` part is internal, it should be _fine_?

like

``` Kotlin 
operator fun get(header: String): HeaderValues {
  return headers[HeaderName(header)]
}

operator fun set(header: String, values: Collection<*>): Request {
  headers[HeaderName(header)] = values.map { it.toString() }
  return this
}
```

so on and so forth.

Reply:
I meant both: I see people do `request.headers.things`. 

I don't mind changing it tho. I do like that the logic of normalisation is actually in `Headers` and not in both `Request` and `Response`. The downside of your proposal is having that on both ends, which could be fine.

Reply:
Yes you are right. I guess the benefit of my proposal is we have to only maintain the code that only matter. 

I am not super against leaving it like this too though. I would say I'll leave the final judgement to you.

Reply:
I think we leave it like this for now, it's not too much work to change later -- which we'll probably do anyway to make Fuel completely threadsafe (read: immutables)